File: packages/altimate-engine/src/altimate_engine/connectors/redshift.py
"""Redshift warehouse connector — extends PostgresConnector with IAM auth."""

from __future__ import annotations

from typing import Any

from altimate_engine.connectors.postgres import PostgresConnector


class RedshiftConnector(PostgresConnector):
    """Amazon Redshift connector extending PostgresConnector.

    Supports:
    - Password auth (inherited from PostgresConnector)
    - IAM role-based auth via boto3 temporary credentials
    """

    def __init__(
        self,
        host: str = "",
        port: int = 5439,  # Redshift default port
        database: str = "dev",
        user: str | None = None,
        password: str | None = None,
        connection_string: str | None = None,
        iam_role: str | None = None,
        region: str | None = None,
        cluster_identifier: str | None = None,
        **kwargs,
    ):
        self.iam_role = iam_role
        self.region = region
        self.cluster_identifier = cluster_identifier
        super().__init__(
            host=host,
            port=port,
            database=database,
            user=user,
            password=password,
            connection_string=connection_string,
            **kwargs,
        )

    def connect(self) -> Any:
        if self.iam_role and not self.password:
            self._resolve_iam_credentials()
        return super().connect()

    def _resolve_iam_credentials(self):
        """Get temporary credentials via IAM role assumption."""
        try:
            import boto3
        except ImportError:
            raise ImportError(
                "boto3 not installed. Install with: pip install boto3"
            )

        if not self.cluster_identifier:
            raise ValueError(
                "cluster_identifier is required for IAM authentication. "
                "This is the Redshift cluster ID (not the full endpoint)."
            )

        client = boto3.client(
            "redshift",
            region_name=self.region or "us-east-1",
        )
        response = client.get_cluster_credentials(
            DbUser=self.user or "admin",
            DbName=self.database,
            ClusterIdentifier=self.cluster_identifier,
        )
        self.user = response["DbUser"]
        self.password = response["DbPassword"]

    def list_schemas(self) -> list[str]:
        rows = self.execute(
            "SELECT schema_name FROM svv_all_schemas "
            "WHERE schema_name NOT IN ('information_schema', 'pg_catalog', 'pg_internal')"
        )
        return [row["schema_name"] for row in rows]

    def list_tables(self, schema: str) -> list[dict[str, Any]]:
        rows = self.execute(
            "SELECT table_name, table_type FROM svv_all_tables WHERE schema_name = %s",
            (schema,),
        )
        return [
            {"name": row["table_name"], "type": row.get("table_type", "TABLE")}
            for row in rows
        ]

    def describe_table(self, schema: str, table: str) -> list[dict[str, Any]]:
        rows = self.execute(
            "SELECT column_name, data_type, is_nullable FROM svv_all_columns "
            "WHERE schema_name = %s AND table_name = %s ORDER BY ordinal_position",
            (schema, table),
        )
        return [
            {
                "name": row["column_name"],
                "data_type": row["data_type"],
                "nullable": row.get("is_nullable", "YES") == "YES",
            }
            for row in rows
        ]

File: packages/drivers/src/redshift.ts
/**
 * Redshift driver using the `pg` package (wire-compatible with PostgreSQL).
 * Uses svv_ system views for introspection.
 */

import type { ConnectionConfig, Connector, ConnectorResult, SchemaColumn } from "./types"

export async function connect(config: ConnectionConfig): Promise<Connector> {
  let pg: any
  try {
    pg = await import("pg")
  } catch {
    throw new Error(
      "Redshift driver not installed (uses pg). Run: npm install pg @types/pg",
    )
  }

  const Pool = pg.default?.Pool ?? pg.Pool
  let pool: any

  const connector: Connector = {
    async connect() {
      const poolConfig: Record<string, unknown> = {}

      if (config.connection_string) {
        poolConfig.connectionString = config.connection_string
      } else {
        // Validate password type to prevent cryptic SASL/SCRAM errors
        if (config.password != null && typeof config.password !== "string") {
          throw new Error(
            "Redshift password must be a string. Check your warehouse configuration.",
          )
        }
        poolConfig.host = config.host ?? "127.0.0.1"
        poolConfig.port = config.port ?? 5439 // Redshift default
        poolConfig.database = config.database ?? "dev"
        poolConfig.user = config.user
        poolConfig.password = config.password
        poolConfig.ssl = config.ssl ?? { rejectUnauthorized: false }
      }

      poolConfig.max = 5
      poolConfig.idleTimeoutMillis = 30000
      poolConfig.connectionTimeoutMillis = 10000

      pool = new Pool(poolConfig)
    },

    async execute(sql: string, limit?: number, _binds?: any[]): Promise<ConnectorResult> {
      const client = await pool.connect()
      try {
        const effectiveLimit = limit ?? 1000
        let query = sql
        const isSelectLike = /^\s*(SELECT|WITH|VALUES)\b/i.test(sql)
        if (
          isSelectLike &&
          effectiveLimit &&
          !/\bLIMIT\b/i.test(sql)
        ) {
          query = `${sql.replace(/;\s*$/, "")} LIMIT ${effectiveLimit + 1}`
        }

        const result = await client.query(query)
        const columns = result.fields?.map((f: any) => f.name) ?? []
        const truncated = result.rows.length > effectiveLimit
        const rows = truncated
          ? result.rows.slice(0, effectiveLimit)
          : result.rows

        return {
          columns,
          rows: rows.map((row: any) =>
            columns.map((col: string) => row[col]),
          ),
          row_count: rows.length,
          truncated,
        }
      } finally {
        client.release()
      }
    },

    async listSchemas(): Promise<string[]> {
      const result = await connector.execute(
        `SELECT DISTINCT schemaname
         FROM svv_tables
         WHERE schemaname NOT IN ('pg_catalog', 'information_schema', 'pg_internal')
         ORDER BY schemaname`,
        10000,
      )
      return result.rows.map((r) => r[0] as string)
    },

    async listTables(
      schema: string,
    ): Promise<Array<{ name: string; type: string }>> {
      const client = await pool.connect()
      try {
        const result = await client.query(
          `SELECT tablename, tabletype
           FROM svv_tables
           WHERE schemaname = $1
           ORDER BY tablename`,
          [schema],
        )
        return result.rows.map((r: any) => ({
          name: r.tablename as string,
          type: String(r.tabletype).toLowerCase() === "view" ? "view" : "table",
        }))
      } finally {
        client.release()
      }
    },

    async describeTable(
      schema: string,
      table: string,
    ): Promise<SchemaColumn[]> {
      const client = await pool.connect()
      try {
        const result = await client.query(
          `SELECT columnname, data_type, is_nullable
           FROM svv_columns
           WHERE schemaname = $1
             AND tablename = $2
           ORDER BY ordinal_position`,
          [schema, table],
        )
        return result.rows.map((r: any) => ({
          name: r.columnname as string,
          data_type: r.data_type as string,
          nullable: String(r.is_nullable).toUpperCase() === "YES",
        }))
      } finally {
        client.release()
      }
    },

    async close() {
      if (pool) {
        await pool.end()
        pool = null
      }
    },
  }
  return connector
}
